### PR TITLE
Implement the BatchingProcessor without polling

### PR DESCRIPTION
**Pull request by MrAlias**
Blocked by #5106 (this includes a cherry-pick of that PRs commit)

The batching log processor provides the base functionality where it batches received records to a configured limit and then exports them. It also provides "polling" functionality where record batches beyond certain time-limit are exported. This change implements the former.

`OnEmit` is implemented batch and flush when a full queue is reached. The `ForceFlush` and `Shutdown` methods are implemented to appropriately interact with this functionality.

The actual exporting functionality is stubbed until #5104 and #5105 are merged.

The addition of polling is planned for a follow up PR. See #5093 for how this will be extended to include polling.

Part of #5063

### CODE REVIEW

File: sdk/log/batch.go [hunk 1]
```diff
@@ -43,39 +59,138 @@ func NewBatchingProcessor(exporter Exporter, opts ...BatchingOption) *BatchingPr
...
-	return nil
+	if b.stopped.Swap(true) {
+		return nil
+	}
+
+	resp := make(chan error, 1)
+	b.enqueue(ctx, b.batch.Flush(), resp)
+
+	// Cancel all exports and polling.
+	b.cancel()
```
Comment (pellared) [inline sdk/log/batch.go:121]: I am not sure why we should cancel pending exports. I think we should still give a chance to have them completed until export timeout is not reached.

Comment (MrAlias) [reply sdk/log/batch.go:121]: This cancels the context for the polling. This is not the context used in the last flush.

Comment (pellared) [reply sdk/log/batch.go:121]:
```suggestion
	// Cancel all subsequent exports and polling.
	b.cancel()
```

**Comment by MrAlias** [comment]
Refactoring based on early received feedback from @pellared has restructured the code in a way that it can be broken into smaller PRs. I have merged this back into https://github.com/open-telemetry/opentelemetry-go/pull/5093 and plan to split from there. Closing.

**Pull request closed without merging**